lib/listen/record.rb

```diff
@@ -32,27 +27,20 @@ def unset_path(dir, rel_path)
     def file_data(dir, rel_path)
       root = @paths[dir.to_s]
       dirname, basename = Pathname(rel_path).split.map(&:to_s)
-      if [nil, '', '.'].include? dirname
-        root[basename] ||= {}
-        root[basename].dup
-      else
-        root[dirname] ||= {}
-        root[dirname][basename] ||= {}
-        root[dirname][basename].dup
-      end
+      dirname = '.' if [nil, '', '.'].include? dirname
+      root[dirname] ||= {}
+      root[dirname][basename] ||= {}
+      root[dirname][basename].dup
     end
 
     def dir_entries(dir, rel_path)
-      tree = if [nil, '', '.'].include? rel_path.to_s
-               @paths[dir.to_s]
-             else
-               @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
-               @paths[dir.to_s][rel_path.to_s]
-             end
+      rel_path = '.' if [nil, '', '.'].include? rel_path.to_s
+      @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
+      tree = @paths[dir.to_s][rel_path.to_s]
 
       result = {}
       tree.each do |key, values|
-        # only get data for file entries
+        # only get data for file/dir entries
         result[key] = values.key?(:mtime) ? values : {}
```

Comment:
> Do directory entries have mtime?

I thought about it - doing so would have removed the need for the `'.'` entries (since `dir_entries` would then also return folders). The alternative having the check on `tree` checking not for the `:mtime` key, but checking if the name has a "/" (non-root-object). So the '.' key is a cache to avoid this (which is why just returning tree works).

The problem is: once you start relying on mtime on OSX, you have to probably take into account the same issue files have - that mtime is unreliable on HFS (that's why you need those md5sum checks on OSX). So you'd probably need to do a "directory md5sum" (which is scanning, anyway) - otherwise you may not detect when directories are added or removed quickly.

I am considering a completely rewrite of Listen to deal with all these issues - though thorough tests are needed either way.